File: windows/src/AINotebook.Core/Providers/OpenWebUIChatAdapter.cs
using AINotebook.Core.Ollama;

namespace AINotebook.Core.Providers;

/// <summary>
/// OpenWebUI streaming chat adapter. OpenWebUI aggregates models (local Ollama,
/// cloud backends, functions) behind an OpenAI-shaped API rooted at /api, NOT /v1:
/// POST {base}/api/chat/completions, GET {base}/api/models.
/// Bearer key optional — instances may run with auth disabled.
/// Chat-only: OpenWebUI exposes no OpenAI-compatible embeddings endpoint.
/// Thin wrapper over <see cref="OpenAIStyleWire"/> — the shared request building,
/// status mapping, and SSE loop live there.
/// </summary>
public sealed class OpenWebUIChatAdapter : IChatStreaming
{
    private const string ChatPath = "/api/chat/completions";
    private const string ModelsPath = "/api/models";

    private readonly HttpClient _http;
    private readonly string _baseUrl;
    private readonly string? _apiKey;

    public OpenWebUIChatAdapter(HttpClient http, string baseUrl, string? apiKey = null)
    {
        _http = http;
        _baseUrl = baseUrl.TrimEnd('/');
        _apiKey = apiKey;
    }

    public IAsyncEnumerable<string> StreamAsync(
        string model,
        IReadOnlyList<ChatTurn> messages,
        CancellationToken ct = default)
    {
        var req = OpenAIStyleWire.BuildChatRequest(_baseUrl, ChatPath, _apiKey, model, messages);
        return OpenAIStyleWire.StreamAsync(_http, req, ct);
    }

    /// <summary>
    /// GET {base}/api/models → {"data":[{"id","name",...}]}. Includes every model the
    /// key's user can access. Throws on ANY failure (macOS parity, via the shared
    /// OpenAIStyleWire helper) — 401 → ProviderAuthException, other HTTP failures →
    /// ProviderException, network errors (HttpRequestException) and cancellation
    /// propagate — so the router's TestConnectionAsync can surface a visible error
    /// message instead of a false "success" with zero models.
    /// </summary>
    public static Task<IReadOnlyList<ProviderModelInfo>> ListModelsAsync(
        HttpClient http, string baseUrl, string? apiKey, CancellationToken ct = default)
        => OpenAIStyleWire.ListModelsAsync(http, baseUrl, ModelsPath, apiKey, ct);
}

File: Sources/AINotebookCore/Providers/OpenWebUIChatAdapter.swift
import Foundation

/// OpenWebUI network adapter. OpenWebUI aggregates models (local Ollama,
/// cloud backends, functions) behind an OpenAI-shape API rooted at /api,
/// NOT /v1: POST {base}/api/chat/completions, GET {base}/api/models.
/// Bearer key optional — instances may run with auth disabled. Chat-only:
/// OpenWebUI exposes no OpenAI-compatible embeddings endpoint.
public struct OpenWebUIChatAdapter: ChatStreaming {
    let baseURL: String
    let apiKey: String?
    let session: URLSession

    public init(baseURL: String, apiKey: String?, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.apiKey = apiKey
        self.session = session
    }

    public func stream(model: String, messages: [ChatTurn]) -> AsyncThrowingStream<String, Error> {
        let request: URLRequest
        do {
            request = try ProviderWire.openAIStyleChatRequest(
                base: baseURL, path: "/api/chat/completions",
                apiKey: apiKey, model: model, messages: messages
            )
        } catch {
            return AsyncThrowingStream { $0.finish(throwing: error) }
        }
        return ProviderWire.openAIStyleStream(request: request, session: session)
    }

    public static func listModels(
        baseURL: String, apiKey: String?, session: URLSession = .shared
    ) async throws -> [ProviderModelInfo] {
        try await ProviderWire.listOpenAIStyleModels(
            base: baseURL, path: "/api/models", apiKey: apiKey, session: session
        )
    }
}
